gef.py

```diff
@@ -10513,15 +10517,26 @@ def pagesize(self) -> int:
 
     @property
     def canary(self) -> Optional[Tuple[int, int]]:
-        """Returns a tuple of the canary address and value, read from the auxiliary vector."""
-        auxval = self.auxiliary_vector
-        if not auxval:
-            return None
-        canary_location = auxval["AT_RANDOM"]
-        canary = gef.memory.read_integer(canary_location)
-        canary &= ~0xFF
-        self._canary = (canary, canary_location)
-        return self._canary
+        """Return a tuple of the canary address and value, read from the canonical
+        location if supported by the architecture. Otherwise, read from the auxiliary
+        vector."""
+        try:
+            canary_location = gef.arch.canary_address()
+            canary = gef.memory.read_integer(canary_location)
+        except NotImplementedError:
+            # Fall back to `AT_RANDOM`, which is the original source
+            # of the canary value but not the canonical location
+            auxval = self.auxiliary_vector
```

Comment:
Can you maybe move this whole block into a dedicated function? That'd allow to keep it tested directly using `gdb_test_python_method()` and allow the CI coverage check to pass.